src/field/gen_fieldops.py

```diff
@@ -0,0 +1,211 @@
+for the arithmetic operators, and prints to stdout.
+
+The `Field` class provides some helper functions for determining how to
+pass a variable by reference or pointer, and how to name arguments in
+function signatures. This allows us to push some logic into the
+templates themselves.
+
+"""
+
+
+from __future__ import print_function
+
+try:
+    from builtins import object
+except ImportError:
+    pass
+
+from collections import OrderedDict
+from copy import deepcopy as copy
+import itertools
+
+try:
+    import jinja2
+except ImportError:
+    raise ImportError('Missing Python module "jinja2". See "Field2D/Field3D Arithmetic '
+                      'Operators" in the BOUT++ user manual for more information')
+
+# The arthimetic operators
+# OrderedDict to (try to) ensure consistency between python 2 & 3
+operators = OrderedDict([
+    ('*', 'multiplication'),
+    ('/', 'division'),
+    ('+', 'addition'),
+    ('-', 'subtraction'),
+])
+
+header = """// This file is autogenerated - see gen_fieldops.py
+#include <bout/mesh.hxx>
+#include <field2d.hxx>
+#include <field3d.hxx>
+#include <globals.hxx>
+#include <interpolation.hxx>
+"""
+
+
+class Field(object):
+    """Abstracts over BoutReals and Field2D/3D/Perps
+
+    Provides some helper functions for writing function signatures and
+    passing data
+
+    """
+
+    def __init__(self, field_type, dimensions, name=None):
+        # C++ type of the field, e.g. Field3D
+        self.field_type = field_type
+        # array: dimensions of the field
+        self.dimensions = dimensions
+        # name of this field
+        self.name = name
+
+    @property
+    def passByReference(self):
+        """Returns "Type& name", except if field_type is BoutReal,
+        in which case just returns "Type name"
+
+        """
+        return "{self.field_type}{ref} {self.name}".format(
+            self=self, ref="&" if self.field_type != "BoutReal" else "")
+
+    @property
+    def passBoutRealPointer(self):
+        """Returns "BoutReal* name", except if field_type is BoutReal,
+        in which case just returns "BoutReal name"
+
+        Also adds "__restrict__" attribute if `for_gcc` is True
+        """
+        return "BoutReal {ref}{restrict} {self.name}".format(
+            self=self, ref="*" if self.field_type != "BoutReal" else "",
+            restrict="__restrict__" if for_gcc and self.field_type != "BoutReal" else "")
+
+    @property
+    def getPointerToData(self):
+        """Returns a pointer to the underlying data, e.g. `&Field3D[i]`,
+        except for plain BoutReals"""
+        if self.field_type == 'BoutReal':
+            return self.name
+        return "&{}[i]".format(self.name)
+
+    def getElement(self, data=True):
+        """Get an element in the field, either using C-loop style index (default),
+        or with an `Indices` object (if `data=False`)
+
+        """
+
+        if self.field_type == 'BoutReal':
```

Comment:
This section looks like it's very much tied to the specific template being used. I'm guessing there's not really much option here, but if we're thinking of expanding the templating a little further we might want to think about expanding the flexibility (e.g. adding arguments to allow us to specify the name of the x index, the size of x etc.  which can have sensible default values).

Reply:
This is true. I don't think it would be too hard to make it more generic, but I think we'd need to see the other use-cases first.

Reply:
I was thinking a bit like https://github.com/boutproject/BOUT-dev/pull/818/files#diff-151f2a716d86ab78dc82ba9d24cd456cR184 where you use the defined dimension strings to form the variable names to use (which I think must match exactly what is used here).

Reply:
Ah, ok. I'll see what can be done.

Reply:
It's not a big issue from my view and changing it may just make things more complicated, but I do think that whilst things are mostly separated there are some dependencies in the choice of these variable names spread across the file which might be non-obvious when trying to extend/reuse.